lib/Dancer2/Core/App.pm

```diff
@@ -1694,6 +1710,63 @@ sub response_not_found {
     return $response;
 }
 
+sub uri_for_route {
+    my ( $self, $route_name, $route_params, $query_params, $dont_escape ) = @_;
+    my $route = $self->route_names->{$route_name}
+        or die "Cannot find route named '$route_name'";
+
+    my $string = $route->spec_route;
+    is_regexpref($string)
+        and die "uri_for_route() does not support regexp route paths";
+
+    # Convert splat only to the general purpose structure
+    if ( is_arrayref($route_params) ) {
+        $route_params = { 'splat' => $route_params };
+    }
+
+    # The regexes are taken and altered from:
+    # Dancer2::Core::Route::_build_regexp_from_string.
+
+    # Replace :foo with arg (route parameters)
+    # Not a fan of all this regex play to handle typed parameter -- SX
+    my @params = $string =~ m{:([^/.\?]+)}xmsg;
+
+    foreach my $param (@params) {
+        $param =~ s{^([^\[]+).*}{$1}xms;
+        my $value = $route_params->{$param}
+            or die "Route $route_name uses the parameter '${param}', which was not provided";
+
+        $string =~ s!\Q:$param\E(\[[^\]]+\])?!$value!xmsg;
+    }
+
+    # TODO: Can we cut this down by replacing on the spot?
+    #       I think that will be tricky because we first need all **, then *
+
+    $string =~ s!\Q**\E!(?#megasplat)!g;
+    $string =~ s!\*!(?#splat)!g;
+
+    # TODO: Can we cut this down?
+    my @token_or_splat =
+          $string  =~ /\(\?#((?:mega)?splat)\)/g;
+
+    my $splat_params = $route_params->{'splat'};
+    if ($splat_params && @token_or_splat) {
+        $#{$splat_params} == $#token_or_splat
+            or die 'Mismatch in amount of splat args and splat elements';
+
+        for ( my $i = 0; $i < @{$splat_params}; $i++ ) {
+            if ( is_arrayref($splat_params->[$i]) ){
+                my $megasplat = join '/', @{ $splat_params->[$i] };
+                $string =~ s{\Q(?#megasplat)\E}{$megasplat};
+            } else {
+                $string =~ s{\Q(?#splat)\E}{$splat_params->[$i]};
+            }
+        }
+    }
+
+    return $self->request->uri_for( $string, $query_params );
```

Comment:
As of 2ff2d530, it looks like `$dont_escape` is still unused.  However, an escape-control parameter is documented.  Should that parameter be passed to this `uri_for()` call?

Reply:
Just noticed this myself. :) Fixed!